File: modules/mesh/negotiator.go
package mesh

import (
	"context"
	"fmt"

	"github.com/woyin/OrangeCoding/modules/core"
)

// HandoffMessage is the data payload published when one agent hands off a
// task to another agent.
type HandoffMessage struct {
	From core.AgentId
	To   core.AgentId
	Task string
}

// Negotiator coordinates task handoffs between agents. It validates that both
// the source and target agents exist in the registry before publishing the
// handoff message on the bus.
type Negotiator struct {
	registry *AgentRegistry
	bus      *MessageBus
}

// NewNegotiator creates a Negotiator wired to the given registry and bus.
func NewNegotiator(registry *AgentRegistry, bus *MessageBus) *Negotiator {
	return &Negotiator{
		registry: registry,
		bus:      bus,
	}
}

// Handoff publishes a task handoff message from one agent to another. Both
// agents must be registered in the registry; otherwise an error is returned.
func (n *Negotiator) Handoff(ctx context.Context, fromID, toID core.AgentId, task string) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}

	if _, ok := n.registry.Get(fromID); !ok {
		return fmt.Errorf("handoff: source agent %s not registered", fromID)
	}
	if _, ok := n.registry.Get(toID); !ok {
		return fmt.Errorf("handoff: target agent %s not registered", toID)
	}

	n.bus.Publish("agent.handoff", HandoffMessage{
		From: fromID,
		To:   toID,
		Task: task,
	})
	return nil
}

// BuddyObserver watches for events on the bus and invokes registered handlers.
type BuddyObserver struct {
	bus      *MessageBus
	handlers map[string]func(data interface{})
}

// NewBuddyObserver creates a BuddyObserver that listens on the given bus.
func NewBuddyObserver(bus *MessageBus) *BuddyObserver {
	return &BuddyObserver{
		bus:      bus,
		handlers: make(map[string]func(data interface{})),
	}
}

// Watch subscribes to events of the given type on the bus and calls handler
// whenever a message arrives on that topic.
func (o *BuddyObserver) Watch(eventType string, handler func(data interface{})) {
	o.handlers[eventType] = handler
	o.bus.Subscribe(eventType, func(topic string, data interface{}) {
		handler(data)
	})
}

File: packages/mesh/src/negotiator.ts
/**
 * @module negotiator
 *
 * Task negotiation between mesh peers.
 *
 * When a task arrives, the negotiator:
 * 1. Classifies the task (complexity, type, required skills)
 * 2. Identifies suitable agents based on capabilities
 * 3. Negotiates task acceptance with candidate agents
 * 4. Assigns the task to the best available agent
 *
 * Supports both automatic assignment and human-in-the-loop approval.
 */

import type { AgentId } from "@orangecoding/core";
import type { AgentRegistry } from "./registry.js";
import type { MessageBus } from "./bus.js";

// ---------------------------------------------------------------------------
// HandoffMessage
// ---------------------------------------------------------------------------

/** Data payload published when one agent hands off a task to another agent. */
export interface HandoffMessage {
  from: AgentId;
  to: AgentId;
  task: string;
}

// ---------------------------------------------------------------------------
// Negotiator
// ---------------------------------------------------------------------------

/**
 * Coordinates task handoffs between agents. It validates that both the source
 * and target agents exist in the registry before publishing the handoff message
 * on the bus.
 */
export class Negotiator {
  private registry: AgentRegistry;
  private bus: MessageBus;

  constructor(registry: AgentRegistry, bus: MessageBus) {
    this.registry = registry;
    this.bus = bus;
  }

  /**
   * Handoff publishes a task handoff message from one agent to another. Both
   * agents must be registered in the registry; otherwise an error is returned.
   */
  handoff(fromID: AgentId, toID: AgentId, task: string): void {
    if (!this.registry.get(fromID)) {
      throw new Error(`handoff: source agent ${fromID} not registered`);
    }
    if (!this.registry.get(toID)) {
      throw new Error(`handoff: target agent ${toID} not registered`);
    }

    this.bus.publish("agent.handoff", {
      from: fromID,
      to: toID,
      task,
    } satisfies HandoffMessage);
  }
}

// ---------------------------------------------------------------------------
// BuddyObserver
// ---------------------------------------------------------------------------

/** Watches for events on the bus and invokes registered handlers. */
export class BuddyObserver {
  private bus: MessageBus;
  private handlers = new Map<string, (data: unknown) => void>();

  constructor(bus: MessageBus) {
    this.bus = bus;
  }

  /**
   * Watch subscribes to events of the given type on the bus and calls handler
   * whenever a message arrives on that topic.
   */
  watch(eventType: string, handler: (data: unknown) => void): void {
    this.handlers.set(eventType, handler);
    this.bus.subscribe(eventType, (_topic: string, data: unknown) => {
      handler(data);
    });
  }
}
